### PR TITLE
Feature/remove compatibility mode

### LINKED ISSUE #798: Package revisions
Most package managers have a concept of package revision, i.e. additional version number that reflects changes in packaging scripts or applied patches when "main" version number of packaged software remains the same.

It would be great if Conan added support for revisions too. This will make package updates more trasparent ("updated from vX.Y.Z-r1 to vX.Y.Z-r2"). Also there could be a policy that "stable" channel can never change conanfile and binaries without bumping revision, to prevent accidental changes in packages used in CI with manifest verification.

It would be great if it was possible to keep binary packages for previous revisions so that CI system with manifests checking does not get broken in case new revision is uploaded without committing new reference manifests.

It was previously briefly discussed at https://github.com/conan-io/conan/issues/480#issuecomment-247545547

### LINKED ISSUE #4352: Revision timestamp update tests
Add tests to verify:

- If the client uploads a recipe/package to a remote, the timestamp of the revision is updated (set from None to something). If the client uploads the same recipe/package to another remote the timestamp should be kept, because the remote in the registry is not updated.

- When the recipe/package is retrieved from one remote, the registry is updated and indeed the timestamp of the revision.

### PR DESCRIPTION
**Pull request by lasote**
Changelog: Feature: Removed old "compatibility" mode of revisions.
Changelog: Bugfix: Fixed several bugs related to revisions.
Changelog: Bugfix: Fixed several bugs related to the package `metadata.json`

Close #4352
Close #798

Docs: PENDING (Search revisions endpoints, What else? should revisions start to have something more than the wiki page? Maybe still a bit early)

CHANGES:

- The CI will run only two stages, "with revisions", "without revisions"
- Different testing env var (`TESTING_REVISIONS_ENABLED`) to change the TestClient default (revisions on/off) than the real one that is `CONAN_CLIENT_REVISIONS_ENABLED`. Otherwise it was very difficul to mix the on/off operations in the tests.
- Metadata is not lazy created by default, it hide errors like "conan copy" not generating the destination metadata files. 
- Metadata is not returning "0" by default for the revisions. It was also error prone. The `None` value can happen if no one sets the revision.
- Decouple "revisions" mechanism from the "SCM" feature, now the revision is extracted directly from the repository using the SCM helpers
- Search and remove needed several changes to work with the revisions.
- Rewritten tests with a new test framework

PENDING:
- [x] Migration of cache to generate the `metadata.json` file
- [ ] Update wiki page about revisions

FUTURE FEATURES TO IMPLEMENT:

- [ ] `revision_mode="scm/hash"` so someone could choose to use the hash (recipes of a third party at the same repository)

@TAGS: svn, slow
@REVISIONS: 1

### CODE REVIEW

File: conans/client/remover.py [hunk 1]
```diff
@@ -96,10 +99,15 @@ def _remote_remove(self, ref, package_ids, remote):
...
+    @staticmethod
+    def _message_removing_editable(ref):
+        return "Package '{r}' is installed as editable, unlink it first using " \
+               "command 'conan link {r} --remove'".format(r=ref)
+
     def _local_remove(self, ref, src, build_ids, package_ids):
         if self._cache.installed_as_editable(ref):
-            raise ConanException("Package '{r}' is installed as editable, unlink it first using "
-                                 "command 'conan link {r} --remove'".format(r=ref))
+            self._user_io.out.warn(self._message_removing_editable(ref))
```
Comment (memsharded) [inline conans/client/remover.py]: Why converting this error to warning? Seems unrelated to this PR.

File: conans/client/remover.py [hunk 2]
```diff
@@ -132,14 +143,48 @@ def remove(self, pattern, remote_name, src=None, build_ids=None, package_ids_fil
...
+        if input_ref and package_ids_filter and not input_ref.revision:
+            for package_id in package_ids_filter:
+                if "#" in package_id:
+                    raise ConanException("Specify a recipe revision if you specify a package "
+                                         "revision")
+
         if remote_name:
             remote = self._registry.remotes.get(remote_name)
-            refs = self._remote_manager.search_recipes(remote, pattern)
+            if input_ref:
```
Comment (memsharded) [inline conans/client/remover.py]: All this code here might be repeated from the "search", please check.

File: conans/model/scm.py [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import json
 import os
 
+from future.moves import subprocess
```
Comment (memsharded) [inline conans/model/scm.py]: Why ``from future``? So far importing subprocess hasn't been an issue.

File: conans/search/search.py [hunk 4]
```diff
@@ -118,35 +134,34 @@ def search_packages(cache, ref, query):
                            settings: {os: Windows}}}
     param ref: ConanFileReference object
     """
-    if not os.path.exists(cache.conan(ref)):
-        raise NotFoundException("Recipe not found: %s" % str(ref))
+    if not os.path.exists(cache.conan(ref)) or (
+           ref.revision and cache.package_layout(ref).recipe_revision()[0] != ref.revision):
```
Comment (memsharded) [inline conans/search/search.py]: ``cache.package_layout(ref).recipe_revision()[0]``? Why the ``[0]``?

File: conans/test/functional/command/search_test.py [hunk 5]
```diff
@@ -175,6 +177,24 @@ def setUp(self):
...
+        def create_metadata(folder, pids):
+            metadata = PackageMetadata()
+            metadata.recipe.revision = DEFAULT_REVISION_V1
+            for pid in pids:
+                metadata.packages[pid].revision = DEFAULT_REVISION_V1
+            save(os.path.join(self.client.cache.store, folder, "metadata.json"), metadata.dumps())
+
+        create_metadata(root_folder1, ["WindowsPackageSHA", "PlatformIndependantSHA",
+                                       "LinuxPackageSHA"])
+        create_metadata(root_folder11, ["WindowsPackageSHA"])
```
Comment (memsharded) [inline conans/test/functional/command/search_test.py]: These changes in this tests, makes me wonder if it is not going to break all users cache that won't have a metadata.json file in their packages yet.

File: conans/test/integration/install_update_test.py [hunk 6]
```diff
@@ -277,9 +277,17 @@ def package(self):
...
+
+        if client.revisions_enabled:
+            # The binary package is not updated but downloaded, because the local one we have
+            # belongs to a different revision and it is removed
+            self.assertIn("Pkg/0.1@lasote/channel:%s - Download" % NO_SETTINGS_PACKAGE_ID,
+                          client.out)
+        else:
+            self.assertIn("Pkg/0.1@lasote/channel:%s - Update" % NO_SETTINGS_PACKAGE_ID,
+                          client.out)
+            self.assertIn("Pkg/0.1@lasote/channel: Retrieving package %s" % NO_SETTINGS_PACKAGE_ID,
```
Comment (memsharded) [inline conans/test/integration/install_update_test.py]: Probably this should be the same "Retrieving package" check both for revisions and no revisions

File: conans/client/remover.py (hunk 1 above)
Comment (lasote) [reply conans/client/remover.py]: Uou, maybe something wrong in a conflict

File: conans/model/scm.py (hunk 3 above)
Comment (lasote) [reply conans/model/scm.py]: IDE autocompleting, I'll fix it

File: conans/search/search.py (hunk 4 above)
Comment (lasote) [reply conans/search/search.py]: It returns a tuple of (revision, revision_time)

File: conans/test/functional/command/search_test.py (hunk 5 above)
Comment (lasote) [reply conans/test/functional/command/search_test.py]: Yes, it will. That's why I wrote there is a pending migration needed in this PR.

File: conans/client/remover.py (hunk 1 above)
Comment (lasote) [reply conans/client/remover.py]: Now I remember, I also completed the remove with editables test. If you are removing by pattern, you probably shouldn't fail if some matching reference is in editable, you want to ignore it, and delete the rest.

File: conans/test/integration/revisions_test.py [hunk 7]
```diff
@@ -1,1062 +1,1452 @@
...
+        self.assertIsNone(rrev_time)  # Is None because it has not been downloaded
+
+        prev, prev_time = client.package_revision(pref)
+        self.assertIsNotNone(prev)
+        self.assertIsNotNone(prev_time)
+        # Local package revision time is the pref2
+        self.assertEquals(prev_time, prev2_time_remote)
+
+    @parameterized.expand([(True,), (False,)])
+    def test_revision_miss_match_packages_in_local(self, v1):
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]:
```suggestion
    def test_revision_mismatch_packages_in_local(self, v1):
```

File: conans/test/integration/revisions_test.py [hunk 8]
```diff
@@ -1,1062 +1,1452 @@
...
+            client.run(command)
+            self.assertIn("{} - Cache".format(pref), client.out)
+        else:
+            client.run(command, assert_error=True)
+            self.assertIn("The package {} doesn't belong to the installed "
+                          "recipe".format(pref), client.out)
+            self.assertIn("ERROR: Missing prebuilt package for '{}'".format(self.ref), client.out)
+
+    @parameterized.expand([(True,), (False,)])
+    def test_revision_install_explicit_miss_match_rrev(self, v1):
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]:
```suggestion
    def test_revision_install_explicit_mismatch_rrev(self, v1):
```

File: conans/test/integration/revisions_test.py [hunk 9]
```diff
@@ -1,1062 +1,1452 @@
...
+            new_client = TurboTestClient(revisions_enabled=True, servers=self.servers)
+            pref = new_client.create(self.ref, conanfile=GenConanfile().with_build_msg("Rev2"))
+            new_client.upload_all(self.ref)
+
+            # Repeat the install --update pointing to the new reference
+            client.run("install {} --update".format(pref.ref.full_repr()))
+            self.assertIn("{} from 'default' - Downloaded".format(self.ref), client.out)
+
+    @parameterized.expand([(True,), (False,)])
+    def test_revision_miss_match_packages_remote(self, v1):
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]:
```suggestion
    def test_revision_mismatch_packages_remote(self, v1):
```

File: conans/test/integration/revisions_test.py [hunk 10]
```diff
@@ -1,1062 +1,1452 @@
...
+        self.c_v2.create(self.ref)
+        self.c_v2.upload_all(self.ref)
+
+        client = self.c_v1 if v1 else self.c_v2
+        client.remove_all()
+        client.export(self.ref, conanfile=GenConanfile().with_build_msg("REV2"))
+        command = "install {}".format(self.ref)
+
+        if v1:
+            client.run(command)
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]: Missing assert after this

File: conans/test/integration/revisions_test.py [hunk 11]
```diff
@@ -1,1062 +1,1452 @@
...
+
+        self.assertNotEqual(rev2, rev)
+        self.assertNotEqual(prev2, prev)
+
+        self.assertIsNotNone(rev2)
+        self.assertIsNotNone(prev2)
+        self.assertIsNone(prev_time2)
+
+    def test_new_exported_revision_always_clear_time(self):
+
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]:
```suggestion
```

File: conans/test/integration/revisions_test.py [hunk 12]
```diff
@@ -1,1062 +1,1452 @@
...
+        client = self.c_v1 if v1 else self.c_v2
+        client.run("remove {} -r default --outdated -f".format(self.ref))
+
+        # The Windows package is not there anymore
+        data = self.c_v1.search(self.ref, remote="default", args="--outdated")
+        oss = [p["settings"]["os"] for p in data["results"][0]["items"][0]["packages"]]
+        self.assertEquals([], oss)
+
+        # But the Linux package is there, not outdated
+        data = self.
```
Comment (danimtb) [inline conans/test/integration/revisions_test.py]: I know that programatically it makes sense but the result could be a bit weird for users :weary:

Comment (danimtb) [inline conans/test/integration/revisions_test.py]: This is one of the reasons to delete packages not matching the recipe revision in export time

File: conans/client/cmd/export_pkg.py [hunk 13]
```diff
@@ -43,7 +44,12 @@ def export_pkg(cache, graph_manager, hook_manager, recorder, output,
         packager.export_pkg(conanfile, package_id, package_folder, dest_package_folder,
                             hook_manager, conan_file_path, ref)
     else:
-        packager.create_package(conanfile, package_id, source_folder, build_folder,
-                                dest_package_folder, install_folder, hook_manager, conan_file_path,
-                                ref, local=True)
+        with cache.package_layout(ref).update_metadata() as metadata:
+            packager.create_package(conanfile, package_id, source_folder, build_folder,
+                                    dest_package_folder, install_folder, hook_manager, conan_file_path,
+                                    ref, local=True)
```
Comment (jgsogo) [inline conans/client/cmd/export_pkg.py]: This should be outside the `with ... metadata:` block, two reasons: 1) it is not used, and 2) there are _hooks_ been called from inside that function and we are considering passing the metadata to hooks, we would lose any changes/additions a hook could do to the `metadata` 👀

File: conans/client/cmd/search.py [hunk 14]
```diff
@@ -57,13 +62,19 @@ def search_packages(self, ref=None, remote_name=None, query=None, outdated=False
...
             recipe_hash = None
 
-        if outdated and recipe_hash:
+        if outdated:
             ordered_packages = filter_outdated(ordered_packages, recipe_hash)
+        elif self._cache.revisions_enabled:
+            # With revisions, by default filter the packages not belonging to the recipe
+            # unless outdated is specified.
+            metadata = self._cache.package_layout(ref).load_metadata()
+            ordered_packages = filter_by_revision(metadata, ordered_packages)
```
Comment (jgsogo) [inline conans/client/cmd/search.py]: I can have `outdated=True` (it is passed from the command line) and _revisions enabled_... 

I think that the `filter_by_revision` should be done always if _revisions enabled_ (we can have packages in the cache corresponding to older revisions of the recipe) and then we should filter the outdated ones if requested by the user (although it might not make sense with the revisions enabled).

Said that it is exactly what it is implemented 🤦‍♂️ ... unless the user forces the revision, or several revisions were created using the same commit (uncommitted changes).

File: conans/client/conan_api.py [hunk 15]
```diff
@@ -153,7 +153,8 @@ def instance_remote_manager(requester, cache, user_io, hook_manager):
 
         # To handle remote connections
         put_headers = cache.read_put_headers()
-        rest_api_client = RestApiClient(user_io.out, requester=requester,
+        rest_api_client = RestApiClient(user_io.out, requester,
+                                        block_v2=not cache.revisions_enabled,
```
Comment (jgsogo) [inline conans/client/conan_api.py]: Rename the parameter name to `revisions_enabled`, more consistent.

File: conans/client/graph/graph_binaries.py [hunk 16]
```diff
@@ -100,6 +84,14 @@ def _evaluate_node(self, node, build_mode, update, evaluated_nodes, remote_name)
                 assert node.recipe != RECIPE_EDITABLE, "Editable package cannot be dirty"
                 rmdir(package_folder)  # Do not remove if it is EDITABLE
 
+            if self._cache.revisions_enabled:
+                metadata = self._cache.package_layout(pref.ref).load_metadata()
+                rec_rev = metadata.packages[pref.id].recipe_revision
+                if rec_rev and rec_rev != node.ref.revision:
+                    output.warn("The package {} doesn't belong "
+                                "to the installed recipe revision, removing folder".format(pref))
+                    rmdir(package_folder)
```
Comment (jgsogo) [inline conans/client/graph/graph_binaries.py]: Consider adding a guard for the `RECIPE_EDITABLE` case. I don't want this code to remove the local folder of my project.

File: conans/client/migrations.py [hunk 17]
```diff
@@ -132,6 +143,43 @@ def _make_migrations(self, old_version):
...
+    for ref in refs:
+        try:
+            folder = cache.export(ref)
+            manifest = FileTreeManifest.load(folder)
+            metadata_path = os.path.join(cache.conan(ref), PACKAGE_METADATA)
+            if not os.path.exists(metadata_path):
+                out.info("Creating {} for {}".format(PACKAGE_METADATA, ref))
+                prefs = _get_prefs(cache, ref)
+                metadata = PackageMetadata()
+                metadata.recipe.revision = manifest.summary_hash
```
Comment (jgsogo) [inline conans/client/migrations.py]: Maybe this revision could have a proper name, as it will have nothing to do with the following ones (those will be the commit hash). A name saying that is was generated from an old Conan version at a certain timestamp.

File: conans/client/rest/rest_client.py [hunk 18]
```diff
@@ -12,7 +12,7 @@ class RestApiClient(object):
         Rest Api Client for handle remote.
     """
 
-    def __init__(self, output, requester, put_headers=None):
+    def __init__(self, output, requester, block_v2, put_headers=None):
```
Comment (jgsogo) [inline conans/client/rest/rest_client.py]: Please, express it as `revisions_enabled`, it will add more consistence across the codebase.

File: conans/client/cmd/search.py (hunk 14 above)
Comment (jgsogo) [reply conans/client/cmd/search.py]: Not really sure about this after looking at the `filter_by_revision` function

File: conans/test/functional/command/remove_test.py [hunk 19]
```diff
@@ -72,7 +75,7 @@ def remove_outdated_test(self):
         test_server = TestServer(users={"lasote": "password"})  # exported users and passwords
         servers = {"default": test_server}
         client = TestClient(servers=servers, users={"default": [("lasote", "password")]})
-        if client.revisions:
+        if client.revisions_enabled:
```
[the remark this answers is not in the document]
Comment (jgsogo) [reply conans/test/functional/command/remove_test.py]: 'TESTING_REVISIONS_ENABLED' !

File: conans/client/migrations.py [hunk 20]
```diff
@@ -132,6 +143,43 @@ def _make_migrations(self, old_version):
...
+                for pref in prefs:
+                    pmanifest = FileTreeManifest.load(cache.package(pref))
+                    metadata.packages[pref.id].revision = pmanifest.summary_hash
+                    metadata.packages[pref.id].recipe_revision = metadata.recipe.revision
+                save(metadata_path, metadata.dumps())
+        except Exception as e:
+            raise ConanException("Something went wrong while generating the metadata.json files "
+                                 "in the cache, please try to fix the issue or wipe the cache: {}"
+                                 ":{}".format(ref, e))
+    out.warn("Migration: Generating missing metadata files finished OK!\n")
```
Comment (memsharded) [inline conans/client/migrations.py]: ``out.success``?

File: conans/client/cmd/export.py [hunk 21]
```diff
@@ -208,6 +206,31 @@ def _replace_scm_data_in_conanfile(conanfile_path, scm_data):
     save(conanfile_path, content)
 
 
+def _detect_scm_revision(path):
+    repo_type = detect_repo_type(path)
+    if not repo_type:
+        return None, None
+
+    repo_obj = {"git": Git, "svn": SVN}.get(repo_type)(path)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/cmd/export.py]: That should be another feature, right?

[the remark this answers is not in the document]
Comment (danimtb) [reply conans/client/cmd/export.py]: Sure!

File: conans/client/conf/__init__.py [hunk 22]
```diff
@@ -338,6 +338,19 @@ def request_timeout(self):
         except ConanException:
             return None
 
+    @property
+    def revisions_enabled(self):
+        try:
+            revisions_enabled = get_env("CONAN_CLIENT_REVISIONS_ENABLED")
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/conf/__init__.py]: Because that exactly what I'm trying to avoid. If I set the env var and then I forget about the config entry, several `TestClients` running at the same time will mix the activation of revisions. This is a general issue of that env var management. But in this case is harder.

File: conans/test/integration/revisions_test.py (hunk 12 above)
Comment (lasote) [reply conans/test/integration/revisions_test.py]: That would be breaking when no revision in the client (this test)

File: conans/client/conf/__init__.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (danimtb) [reply conans/client/conf/__init__.py]: ok, so you could enable the configuration of each client instead of using env vars. That way each client will work with or without revisions independently and we will keep the same code as any other config entry

File: conans/client/cmd/copy.py [hunk 23]
```diff
@@ -72,3 +79,11 @@ def package_copy(src_ref, user_channel, package_ids, paths, user_io, short_paths
             rmdir(package_path_dest)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/cmd/copy.py]: I'll move the metadata assignment after the confirmation. About if we should let the user choose or not, I think it is ok but it is not definitely the scope of this PR.

File: conans/client/cmd/copy.py [hunk 24]
```diff
@@ -39,6 +40,9 @@ def package_copy(src_ref, user_channel, package_ids, paths, user_io, short_paths
     dest_ref = ConanFileReference.loads("%s/%s@%s" % (src_ref.name,
                                                       src_ref.version,
                                                       user_channel))
+    # Generate metadata
+    old_metadata = paths.package_layout(src_ref).load_metadata()
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/cmd/copy.py]: ok

File: conans/client/cmd/copy.py [hunk 25]
```diff
@@ -72,3 +79,11 @@ def package_copy(src_ref, user_channel, package_ids, paths, user_io, short_paths
...
         shutil.copytree(package_path_origin, package_path_dest, symlinks=True)
         user_io.out.info("Copied %s to %s" % (str(package_id), str(dest_ref)))
+
+    # Generate the metadata
+    with paths.package_layout(dest_ref).update_metadata() as metadata:
+        metadata.recipe.revision = old_metadata.recipe.revision
+        for package_id, (revision, recipe_revision) in package_revisions.items():
+            metadata.packages[package_id].revision = revision
+            metadata.packages[package_id].recipe_revision = recipe_revision
+
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/cmd/copy.py]: do it. Low priority. The copy is very very low usage.

File: conans/client/cmd/search.py (hunk 14 above)
Comment (lasote) [reply conans/client/cmd/search.py]: I think this is correct.

File: conans/client/migrations.py (hunk 17 above)
Comment (lasote) [reply conans/client/migrations.py]: It is the regular revision when you don't have `scm` or when, in the future, you can choose if hash or not hash. What do you want exactly to name here? This is just to create a default. The hash as the revision is a good start for old packages.

File: conans/test/functional/command/search_test.py [hunk 26]
```diff
@@ -1173,6 +1210,13 @@ class Test(ConanFile):
 
         # LOCAL CACHE CHECKS
         client.run("search %s --revisions" % full_ref.format(rrev=first_rrev))
+
+        # The time is not updated locally until we install the package
+        self.assertIn("%s (No time)" % first_prev, client.out)
+        client.run("remove lib/1.0@user/testing -f")
+        client.run("install lib/1.0@user/testing")
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/functional/command/search_test.py]: No, because there are no updates. I will add a test to check that it doesn't help.

File: conans/client/cmd/search.py (hunk 14 above)
Comment (jgsogo) [reply conans/client/cmd/search.py]: me too, after reviewing the tests I think it works as expected.

File: conans/client/migrations.py (hunk 17 above)
Comment (jgsogo) [reply conans/client/migrations.py]: If all the revisions but this one will be named after the `scm` sha (I choose to use the default behavior with a git repo) then I will end up with a list of revisions corresponding to commits in my repository; I can clone any of them but this one, and it looks like a commit sha but it isn't... If we name it "Conan migration - <timestamp>" I'm sure it is not a commit and I won't try to clone it (and wonder why I lost that commit in my repo).

File: conans/test/integration/revisions_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: I'll take the @danimtb recommendation to remove non-matching revision binary packages, but only when revisions enabled, otherwise it would be breaking.
So in this test, there is nothing to change because the exported stuff is done with v1. I'll add an additional test to test the new behavior.

[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: The new functions to auto-detect the repository type to be able to capture the revision is new, and that is not tested elsewhere.

[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: I don't get it.

**Comment by jgsogo** [comment]
⚠️  The migration fails with packages installed in editable mode:

```
ERROR: Something went wrong while generating the metadata.json files in the cache, please try to fix the issue or wipe the cache: lib/version@user/channel:Operation not allowed on a package installed as editable
```

File: conans/test/integration/revisions_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jgsogo) [reply conans/test/integration/revisions_test.py]: I'm suggesting to change:

```python
        # Delete the package revisions
        self.c_v2.run("remove {} -p {}#{} -r default -f".format(pref3.ref.full_repr(),
                                                                pref3.id, pref3.revision))
        self.c_v2.run("remove {} -p {}#{} -r default -f".format(pref2.ref.full_repr(),
                                                                pref2.id, pref2.revision))
        self.c_v2.run("remove {} -p {}#{} -r default -f".format(pref1.ref.full_repr(),
                                                                pref1.id, pref1.revision))
```

by:

```python
# Delete the package revisions (all of them have the same ref#rev and id)
command = "remove {} -p {}#{{}} -r default -f".format(pref3.ref.full_repr(), pref3.id)
self.c_v2.run(command.format(pref3.revision))
self.c_v2.run(command.format(pref2.revision))
self.c_v2.run(command.format(pref1.revision))
```

[the remark this answers is not in the document]
Comment (jgsogo) [reply conans/test/integration/revisions_test.py]: Add tests for that function then, should be smaller (unit tests) and quicker

File: conans/test/integration/revisions_test.py [hunk 27]
```diff
@@ -1,1062 +1,1452 @@
...
+        new RREV. If a client installs without specifying -r remote1, it WONT iterate
+        remote2, because it is associated in the registry and have it in the cache. Unless we
+        specify the -r remote2"""
+
+        conanfile = GenConanfile().with_package_file("file.txt", env_var="MY_VAR")
+        with environment_append({"MY_VAR": "1"}):
+            pref = self.c_v2.create(self.ref, conanfile=conanfile)
+        self.c_v2.upload_all(self.ref, remote="default")
+
+        time.sleep(1)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: Is not about the revision time but the manifest timestamp. we would need to mock other things, I'm not saying it is not possible but not now.

File: conans/test/integration/revisions_test.py [hunk 28]
```diff
@@ -1,1062 +1,1452 @@
...
+        # Install, it wont resolve the remote2 because it is in the registry, it will use the cache
+        self.c_v2.run("install {} --update".format(self.ref))
+        self.assertIn("lib/1.0@conan/testing:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - "
+                      "Cache".format(pref.id), self.c_v2.out)
+
+        # If we force remote2, it will find an update
+        self.c_v2.run("install {} --update -r remote2".format(self.ref))
+        self.assertIn("{} - Update".format(pref), self.c_v2.out)
+        self.assertIn("Retrieving package {} from remote 'remote2' ".format(pref.id),
+                      self.c_v2.out)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: Actually, it doesn't change the remote with an update. Not sure if it was intended or not, but not this PR purpose. I will leave the assert and open an issue.

File: conans/test/integration/revisions_test.py [hunk 29]
```diff
@@ -1,1062 +1,1452 @@
...
+
+    def test_alias_to_a_rrev(self):
+        """ If an alias points to a RREV, it resolved that RREV and no other"""
+
+        # Upload one revision
+        pref = self.c_v2.create(self.ref)
+        self.c_v2.upload_all(self.ref)
+
+        # Upload other revision
+        self.c_v2.create(self.ref, conanfile=GenConanfile().with_build_msg("Build Rev 2"))
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: Good catch, it is not working actually.

File: conans/test/integration/revisions_test.py [hunk 30]
```diff
@@ -1,1062 +1,1452 @@
...
+        client = TurboTestClient(revisions_enabled=True, servers={"default": self.server})
+        client2 = TurboTestClient(revisions_enabled=True, servers={"default": self.server})
 
-        self.client.run("create . %s" % str(ref))
-        # The revision for the package is also there
-        prev, tprev = self.client.cache.package_layout(ref).package_revision(pref)
+        client.create(self.ref)
+        client.upload_all(self.ref)
+
+        time.sleep(1)  # Wait a second, to be considered an update
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: same. Would be nice to provide an easy mock for the manifest, but cannot do it now.

File: conans/test/integration/revisions_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: Nope. wrong comment

[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/integration/revisions_test.py]: ah! nice.

File: conans/client/migrations.py (hunk 17 above)
Comment (lasote) [reply conans/client/migrations.py]: tiny improvement...

**Comment by lasote** [comment]
Great reviews team.
